.gitbook/contents/hc/4_deploy-smart-contract.md

```diff
@@ -1,49 +1,119 @@
+
+```solidity
 // These contracts are one example each
 ret[0] = address(new AuctionFactory(ha1Addr));
 ret[1] = address(new TestCaptcha(ha1Addr));
 ret[2] = address(new TestCounter(ha1Addr));
 ret[3] = address(new RainfallInsurance(ha1Addr));
 ret[4] = address(new SportsBetting(ha1Addr));
+```
 
-vm.stopBroadcast();
+Within `deploy-local.py` the Forge scripts are invoked like this:
+
+```python
+  args = ["forge", "script", "--json", "--broadcast", "--silent"]
+  args.append ("--rpc-url=http://127.0.0.1:9545")
+  args.append("hc_scripts/LocalDeploy.s.sol")
+  cmd_env = os.environ.copy()
+  cmd_env['PRIVATE_KEY'] = deploy_key
+  cmd_env['HC_SYS_OWNER'] = env_vars['HC_SYS_OWNER']
+  cmd_env['DEPLOY_ADDR'] = deploy_addr
+  cmd_env['DEPLOY_SALT'] = cli_args.deploy_salt  # Update to force redeployment
 ```
 
+The Forge scripts may be run manually by setting the corresponding environment variables
+and then using the `forge script` command as documented at https://book.getfoundry.sh/reference/forge/forge-script
+
 {% hint style="info" %}
-In the above example we deploy with Foundry, but feel free to deploy with whatever tooling you wish.
+We deploy with Foundry, but you may deploy with whatever tooling you wish.
 {% endhint %}
 
+Further system setup is performed through the web3.py API, first by connecting to our L1 and L2 networks:
+
+```python
+l1 = Web3(Web3.HTTPProvider("http://127.0.0.1:8545"))
+assert (l1.is_connected)
+l1.middleware_onion.inject(geth_poa_middleware, layer=0)
+
+w3 = Web3(Web3.HTTPProvider(env_vars['NODE_HTTP']))
+assert (w3.is_connected)
+```
+
+Next, we access the contracts using a `loadContract()` function, such as
+for our `HybridAccount` contract:
+
+```python
+HA = loadContract(w3, "HybridAccount", path_prefix+"samples/HybridAccount.sol")
+```
+
+Here `path_prefix` indicates where the contract's source code is located. The current deployer script
+will compile the Solidity code itself to determine the contract ABI, rather than requiring a JSON file
+from a previous deployment.
+
+After loading the contract, we can interact with it like this:
+
+```python
+def permitCaller(acct, caller):
+    if not acct.functions.PermittedCallers(caller).call():
+        print("Permit caller {} on {}".format(caller, acct.address))
+        tx = acct.functions.PermitCaller(caller, True).build_transaction({
+            'nonce': w3.eth.get_transaction_count(env_vars['OC_OWNER']),
+            'from': env_vars['OC_OWNER'],
+            'gas': 210000,
+            'chainId': HC_CHAIN,
+        })
+        signAndSubmit(w3, tx, env_vars['OC_PRIVKEY'])
+
+# ...
+
+for a in example_addrs:
+  permitCaller(HA, a)
+
+```
+
+Here we are performing a sequence of transactions to add individual example contract addresses to the
+list of contracts which are permitted to use our HybridAccount. There is a check so that this operation
+is only performed for contracts which have not yet been registered.
```

Comment:
```suggestion
Here we perform a sequence of transactions to add individual example contract addresses to the
list of contracts which are permitted to use our `HybridAccount`. There is a check so that this operation
is only performed for contracts which have not yet been registered.
```